src/db.rs

```diff
@@ -1588,6 +1651,66 @@ impl DB {
     pub fn latest_sequence_number(&self) -> u64 {
         unsafe { ffi::rocksdb_get_latest_sequence_number(self.inner) }
     }
+
+    /// Iterate over batches of write operations since a given sequence.
+    ///
+    /// Produce an iterator that will provide the batches of write operations
+    /// that have occurred since the given sequence (see
+    /// `latest_sequence_number()`). Use the provided iterator to retrieve each
+    /// (`u64`, `WriteBatch`) tuple, and then gather the individual puts and
+    /// deletes using the `WriteBatch::iterate()` function.
+    ///
+    /// Calling `get_updates_since()` with a sequence number that is out of
+    /// bounds will return an error.
+    pub fn get_updates_since(&self, seq_number: u64) -> Result<DBWALIterator, Error> {
+        unsafe {
+            // rocksdb_wal_readoptions_t does not appear to have any functions
+            // for creating and destroying it; fortunately we can pass a nullptr
+            // here to get the default behavior
+            let opts: *const ffi::rocksdb_wal_readoptions_t = ptr::null();;
+            let iter = ffi_try!(ffi::rocksdb_get_updates_since(self.inner, seq_number, opts,));
+            Ok(DBWALIterator { inner: iter })
+        }
+    }
+}
+
+/// Receives the puts and deletes of a write batch.
+///
+/// The application must provide an implementation of this trait when
+/// iterating the operations within a `WriteBatch`
+pub trait WriteBatchIterator {
+    /// Called with a key and value that were `put` into the batch.
+    fn put(&mut self, key: Box<[u8]>, value: Box<[u8]>);
+    /// Called with a key that was `delete`d from the batch.
+    fn delete(&mut self, key: Box<[u8]>);
+}
+
+unsafe extern "C" fn writebatch_put_callback(
+    state: *mut c_void,
+    k: *const c_char,
+    klen: usize,
+    v: *const c_char,
+    vlen: usize,
+) {
+    // coerce the raw pointer back into a box, but "leak" it so we prevent
```

Comment:
I do not understand the purpose of leaking box here. Shouldn't it be available for the whole time of `rocksdb_writebatch_iterate` function?

Reply:
Because the boxed callback thing has to be passed through C and back again, we need the raw pointer. But when you call `Box::from_raw()` you get a _managed_ box. Now the callback thing would get freed. On the second call, we crash because of a double-free. To prevent that, we "leak" the box, telling the compiler to "forget" about it. The box is managed in the `WriteBatch.iterate()`, which is why I had the `Box::from_raw()` there, properly disposing of the box when we are completely done with it.